events: Start integrating pydantic types into real code.

showell opened the pull request.

showell committed:
event types: Eliminate id in BaseEvent.

For all of our real code, we don't need it, and for our
test code, we already funnel all of our checks through
the same helper, and that helper can easily check id,
then delete it, and move on.

showell committed:
events: Start using pydantic types (alert words).

showell committed:
attachment events: Use pydantic.

gnprice commented on zerver/models/custom_profile_fields.py:
```diff
@@ -174,9 +186,8 @@ def as_dict(self) -> ProfileDataElementBase:
+            "display_in_profile_summary": self.display_in_profile_summary,
         }
-        if self.display_in_profile_summary:
-            data_as_dict["display_in_profile_summary"] = True
```
This would be an API change. It needs a corresponding update to the docs, and before merge it needs a discussion in #api design.

The commit message says it wasn't documented, but I see it on https://zulip.com/api/register-queue like so:
![image](https://github.com/user-attachments/assets/87cf7c59-95da-4eaf-8a2f-54c637c557e2)

Even when not previously documented, API changes need discussion in #api design. There's not much of the API that's undocumented at this point, but much of what's left is still things that clients are relying on — just as the mobile client was using plenty of the API before Zulip had API docs at all.

showell replied:
I removed the commit here in question.

showell committed:
event types: Use Event as suffix instead of prefix.

showell committed:
default_stream_groups event: Use pydantic.

showell committed:
default_streams event: Use pydantic.

showell committed:
delete_messages event: Use four new pydantic classes.

In passing, I also flipped a conditional to avoid
double negatives. I also removed an unnecessary
variable intialization.

Please take note of the comment in check-schemas.
It's an imperfect approach for now, but we intend
to clean it up in the future when we have more
similar examples to guide us.

showell committed:
draft event tests: Use stricter pydantic type.

It turns out that timestamp always gets sent from
zerver.models.Draft.to_dict(), and the docs agree.

timabbott commented on zerver/lib/event_types.py:
```diff
@@ -42,26 +42,26 @@ class Attachment(BaseModel):
 
-class AttachmentFieldForAttachmentRemoveEvent(BaseModel):
+class IdHolder(BaseModel):
     id: int
```
This could probably use a docstring. I'm not sure why it isn't just a normal `int` annotation?

showell replied:
It's a one-field struct that holds an int.

showell replied:
It's the difference between `attachment_id: 99` and `attachment: {"id": 99}`.  For some reason we chose to use the latter in our API.  I think `IdHolder` -> `IdStruct` may be an improvement here.  I don't know what to put in the docstring.

timabbott replied:
`IdHolder` seems fine. This seems OK for a docstring.

> A simple object with a single key, an integer ID. Presumably a simple ID was not selected for extensibility or something.

showell replied:
Added a comment with different wording.

timabbott commented on tools/check-schemas:
```diff
@@ -71,6 +71,9 @@ def get_event_checker(event: dict[str, Any]) -> Callable[[str, dict[str, Any]],
     name = name + "Event"
 
+    if name == "DeleteMessageEvent":
+        name = "DeleteStreamMessagesEvent"
```
This seems like an odd construction. It should get a comment at least, but ideally wouldn't need to exist.

showell replied:
The previous version of `DeleteMessageEvent` got split out into four new classes.  We only have one example in our JS fixtures that needs to be verified, and it's the `DeleteStreamMessagesEvent` flavor`.  I chose to use the two-line hack here vs. creating a union type on the Python side.  The full commit puts this more into context.  I'm open to other ideas here.

If you agree that the two-line hack here is the most expedient solution, I will basically just put some version of this PR comment into the code.  I actually think there's probably a better solution here.  Maybe not for this specific problem (it's hard to argue with a simple 2-line hack) but holistically.

timabbott replied:
Ahh, OK -- yeah just add a small comment, we can clean it up later once we've improved the JS fixtures.

showell replied:
Added actually kind of a long comment here, actually, but feel free to edit it down.

showell commented:
I've added comments to the code to address Tim's points above. This is ready for review again.  Thanks for the feedback.
